### PR TITLE
#159965307 Rate an article

### PR DESCRIPTION
**Pull request by micahoriaso**
#### What does this PR do?
- Enable users to rate articles
- The rating scale is 1-5
#### Description of task to be completed?
- Give users the ability to rate articles on a scale of 1 to 5

#### How should this be manually tested?
$ git clone https://github.com/andela/ah-magnificent6.git
$ git checkout ft-rate-article-159965307
$ cd ah-magnificent6
$ cp .env-sample .env [and set the environment variables]
$ python3 -m venv venv
$ source venv/bin/activate
(venv) $ pip install -r requirements.txt
(venv) $ python manage.py migrate
(venv) $ python manage.py runserver
Using [Postman](https://www.getpostman.com/) navigate to 
User registration: `POST {your-server-root}:8000/api/user/signup`
```
{
    "email": "author@user.user",
    "username": "author",
    "password": "good123456"
}
```

User Login: `POST {your-server-root}:8000/api/user/login`
```
{
    "email": "author@user.user",
    "password": "good123456"
}
```
Use the token from the response in the login as the authentication header while creating an article.

screen shot 2018-09-12 at 10 54 16

Create article : POST {your-server-root}:8000/api/articles/
```
{
"title": "My Journey to Andela",
 "description": "This article is about how I joined Andela",
"body": "This is my story to Andela"
}
```

> POST {your-server-root}:8000/api/articles/`<article-slug>`/rate/
with a payload of this format

```
{
"rating": rating_value
}
```
for example 

```
{
"rating": 5
}
```

#### What are the relevant pivotal tracker stories?
[#159965307](https://www.pivotaltracker.com/story/show/159965307)
#### Screenshots
screen shot 2018-09-13 at 09 39 17

### CODE REVIEW

File: authors/apps/articles/renderers.py [hunk 1]
```diff
@@ -5,12 +5,13 @@
 
 
 class ArticleJSONRenderer(JSONRenderer):
+    """JSONRenderClass for formating Article model data into JSON."""
```
Comment (meshack-mbuvi) [inline authors/apps/articles/renderers.py]: Check the spellings for `formatting`

File: authors/apps/articles/tests/test_rate_article.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            {'rating': -4},
+            format="json",
+            **self.headers)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(str(response.data['errors']['rating'][0]),
+                         self.violate_min_value_error_message)
+
+    def test_unsuccessful_rating_with_rate_value_more_than_five(self):
+        """
+        Tests that a user can rate a specific article
```
Comment (meshack-mbuvi) [inline authors/apps/articles/tests/test_rate_article.py]: Remember to update the doc strings for your tests; They seem to be similar even when you have indicated to be testing both positive and negative outcomes.

**Review by meshack-mbuvi — CHANGES_REQUESTED** [review]

File: authors/apps/articles/renderers.py (hunk 1 above)
Comment (micahoriaso) [reply authors/apps/articles/renderers.py]: This has been corrected

File: authors/apps/articles/tests/test_rate_article.py (hunk 2 above)
Comment (micahoriaso) [reply authors/apps/articles/tests/test_rate_article.py]: This has been corrected

**Review by meshack-mbuvi — APPROVED** [review]

**Comment by Bileonaire** [comment]
@micahoriaso Good work on the overwriting users rating once they rate an article more than once. Think of how you can add the delete rating when a user wants to remove their rating completely rather than add a new rating. 
But it is not compulsory, we can do without it for now because we need the ratings so much for the authors to get feedback.

**Comment by micahoriaso** [comment]
> @micahoriaso Good work on the overwriting users rating once they rate an article more than once. Think of how you can add the delete rating when a user wants to remove their rating completely rather than add a new rating.
> But it is not compulsory, we can do without it for now because we need the ratings so much for the authors to get feedback.

Good catch @Bileonaire . I thought about it but I have rarely seen applications where users delete their ratings. That is why I have not included it.

File: authors/apps/articles/models.py [hunk 3]
```diff
@@ -22,6 +24,8 @@ class Meta:
     slug = models.SlugField(unique=True, editable=False, max_length=140)
     favorited = models.NullBooleanField(default=False)
     favoritesCount = models.IntegerField(default=0)
+    rating_average = models.DecimalField(
```
Comment (parseen254) [inline authors/apps/articles/models.py]: @micahoriaso what scale of rating are we using?

Comment (micahoriaso) [reply authors/apps/articles/models.py]: We're rating on a scale of 1-5

**Pull request merged**